### ISSUE TITLE
Snapshot integrity failure error when creating nested resources in a large stack

### ISSUE THREAD

**Issue by aryehb** [pulumi/pulumi#15155]
### What happened?

We have a large stack with around 5000 resources, using local filesystem backend. A good portion of those resources are multiple instances of a nested structure that looks like this (some information has been redacted):

```
urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent::NAME
├─ urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent$gcp:certificatemanager/dnsAuthorization:DnsAuthorization::NAME-1
│  └─ urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent$gcp:certificatemanager/dnsAuthorization:DnsAuthorization$gcp:dns/recordSet:RecordSet::dns-auth-_acme-challenge-NAME-cname-1
├─ urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent$gcp:certificatemanager/dnsAuthorization:DnsAuthorization::NAME-2
│  └─ urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent$gcp:certificatemanager/dnsAuthorization:DnsAuthorization$gcp:dns/recordSet:RecordSet::dns-auth-_acme-challenge-NAME-cname-2
├─ urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent$gcp:certificatemanager/dnsAuthorization:DnsAuthorization::NAME-3
│  └─ urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent$gcp:certificatemanager/dnsAuthorization:DnsAuthorization$gcp:dns/recordSet:RecordSet::dns-auth-_acme-challenge-NAME-cname-3
├─ urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent$gcp:certificatemanager/dnsAuthorization:DnsAuthorization::NAME-4
│  └─ urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent$gcp:certificatemanager/dnsAuthorization:DnsAuthorization$gcp:dns/recordSet:RecordSet::dns-auth-_acme-challenge-NAME-cname-4
├─ urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent$gcp:certificatemanager/dnsAuthorization:DnsAuthorization::NAME-5
│  └─ urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent$gcp:certificatemanager/dnsAuthorization:DnsAuthorization$gcp:dns/recordSet:RecordSet::dns-auth-_acme-challenge-NAME-cname-5
├─ urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent$gcp:certificatemanager/certificate:Certificate::NAME
│  ├─ urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent$gcp:certificatemanager/certificate:Certificate$gcp:certificatemanager/certificateMapEntry:CertificateMapEntry::NAME-1
│  ├─ urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent$gcp:certificatemanager/certificate:Certificate$gcp:certificatemanager/certificateMapEntry:CertificateMapEntry::NAME-2
│  ├─ urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent$gcp:certificatemanager/certificate:Certificate$gcp:certificatemanager/certificateMapEntry:CertificateMapEntry::NAME-3
│  ├─ urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent$gcp:certificatemanager/certificate:Certificate$gcp:certificatemanager/certificateMapEntry:CertificateMapEntry::NAME-4
│  └─ urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent$gcp:certificatemanager/certificate:Certificate$gcp:certificatemanager/certificateMapEntry:CertificateMapEntry::NAME-5
└─ urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent$gcp:dns/managedZone:ManagedZone::managed-zone-NAME
   ├─ urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent$gcp:dns/managedZone:ManagedZone$gcp:dns/recordSet:RecordSet::recordset-NAME-01
   ├─ urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent$gcp:dns/managedZone:ManagedZone$gcp:dns/recordSet:RecordSet::recordset-NAME-02
   ├─ urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent$gcp:dns/managedZone:ManagedZone$gcp:dns/recordSet:RecordSet::recordset-NAME-03
   ├─ urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent$gcp:dns/managedZone:ManagedZone$gcp:dns/recordSet:RecordSet::recordset-NAME-04
   ├─ urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent$gcp:dns/managedZone:ManagedZone$gcp:dns/recordSet:RecordSet::recordset-NAME-05
   ├─ urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent$gcp:dns/managedZone:ManagedZone$gcp:dns/recordSet:RecordSet::recordset-NAME-06
   ├─ urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent$gcp:dns/managedZone:ManagedZone$gcp:dns/recordSet:RecordSet::recordset-NAME-07
   ├─ urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent$gcp:dns/managedZone:ManagedZone$gcp:dns/recordSet:RecordSet::recordset-NAME-08
   ├─ urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent$gcp:dns/managedZone:ManagedZone$gcp:dns/recordSet:RecordSet::recordset-NAME-09
   ├─ urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent$gcp:dns/managedZone:ManagedZone$gcp:dns/recordSet:RecordSet::recordset-NAME-10
   ├─ urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent$gcp:dns/managedZone:ManagedZone$gcp:dns/recordSet:RecordSet::recordset-NAME-11
   └─ urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent$gcp:dns/managedZone:ManagedZone$gcp:dns/recordSet:RecordSet::recordset-NAME-12
```

Recently, when I tried creating another instance of this, the update failed with errors. The `ManagedZone` and `DnsAuthorization`s were created, but the update failed with errors that looked like this:
```
    error: post-step event returned an error: failed to save snapshot: .pulumi/stacks/live.json: snapshot integrity failure; it was already written, but is invalid (backup available at .pulumi/stacks/live.json.bak): child resource urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent$gcp:certificatemanager/dnsAuthorization:DnsAuthorization::NAME-1 refers to missing parent urn:pulumi:live::pulumi-live::ORG:iso-components:IsoComponent::NAME
```

I tried several times to delete the resources from GCP and the Pulumi state and re-run the update, but I get the same results each time.

### Example

N/A

### Output of `pulumi about`

(The full URN list has been redacted. I can provide that privately if necessary)

CLI          
Version      3.94.2
Go Version   go1.21.4
Go Compiler  gc

Plugins
NAME    VERSION
nodejs  unknown

Host     
OS       ubuntu
Version  20.04
Arch     x86_64

This project is written in nodejs: executable='/home/USERNAME/.nvm/versions/node/v18.12.1/bin/node' version='v18.12.1'

Current Stack: live

TYPE                                                               URN
pulumi:pulumi:Stack                                                urn:pulumi:live::pulumi-live::pulumi:pulumi:Stack::pulumi-live-live
pulumi:providers:gcp                                               urn:pulumi:live::pulumi-live::pulumi:providers:gcp::default_6_58_0

OPP TYPE  URN
updating  urn:pulumi:live::pulumi-live::gcp:compute/uRLMap:URLMap::url-map-http

Backend        
Name           USERNAME
URL            file://.
User           USERNAME
Organizations  
Token type     personal

Dependencies:
NAME                   VERSION
@pulumi/cloudflare     5.0.0
@pulumi/gcp            6.58.0
@pulumi/pulumi         3.72.2
@types/lodash          4.14.194
@types/node            18.19.7
lodash                 4.17.21
typescript             4.9.5
@google-cloud/domains  2.2.2

Pulumi locates its logs in /tmp by default

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by justinvp** [comment]
@aryehb, thanks for opening the issue and sorry for the trouble. Do you still get this error using the latest version of the CLI (v3.100.1)? If so, is there any chance you could share a minimal repro? Ideally a pared down and sanitized/redacted version of your program where this still repros. If you'd like to send it privately, you can send it to justin@pulumi.com. Thanks!

**Comment by aryehb** [comment]
@justinvp Thank you. The issue appears to be resolved after updating to v3.101.1.
